klippy/kinematics/deltesian.py

```diff
@@ -0,0 +1,184 @@
+# Code for handling the kinematics of deltesian robots
+#
+# Copyright (C) 2022  Fabrice Gallet <tircown@gmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+import math, logging
+import stepper
+
+# Slow moves once the ratio of tower to XY movement exceeds SLOW_RATIO
+SLOW_RATIO = 3.
+
+# Minimum angle with the horizontal for the arm to not exceed - in degrees
+MIN_ANGLE = 5.
+
+class DeltesianKinematics:
+    def __init__(self, toolhead, config):
+        self.rails = [None] * 3
+        stepper_configs = [config.getsection('stepper_' + s)
+                                    for s in ['left', 'right', 'y']]
+        self.rails[0] = stepper.PrinterRail(
+            stepper_configs[0], need_position_minmax = False)
+        def_pos_es = self.rails[0].get_homing_info().position_endstop
+        self.rails[1] = stepper.PrinterRail(
+            stepper_configs[1], need_position_minmax = False,
+            default_position_endstop = def_pos_es)
+        self.rails[2] = stepper.LookupMultiRail(stepper_configs[2])
+        arm_x = self.arm_x = [None] * 2
+        arm_x[0] = stepper_configs[0].getfloat('arm_x_length', above=0.)
+        arm_x[1] = stepper_configs[1].getfloat('arm_x_length', arm_x[0],
+                                                above=0.)
+        arm = [None] * 2
+        arm[0] = stepper_configs[0].getfloat('arm_length', above=arm_x[0])
+        arm[1] = stepper_configs[1].getfloat('arm_length', arm[0],
+                                              above=arm_x[1])
+        arm2 = self.arm2 = [a**2 for a in arm]
+        self.rails[0].setup_itersolve(
+            'deltesian_stepper_alloc', arm2[0], -arm_x[0])
+        self.rails[1].setup_itersolve(
+            'deltesian_stepper_alloc', arm2[1], arm_x[1])
+        self.rails[2].setup_itersolve('cartesian_stepper_alloc', b'y')
+        for s in self.get_steppers():
+            s.set_trapq(toolhead.get_trapq())
+            toolhead.register_step_generator(s.generate_steps)
+        config.get_printer().register_event_handler(
+            "stepper_enable:motor_off", self._motor_off)
+        self.limits = [(1.0, -1.0)] * 3
+        # X axis limits
+        min_angle = config.getfloat('min_angle', MIN_ANGLE,
+                                    minval=0., maxval=90.)
+        cos_angle = math.cos(math.radians(min_angle))
+        x_kin_min = math.ceil( -min(arm_x[0], cos_angle * arm[1] - arm_x[1]))
+        x_kin_max = math.floor( min(arm_x[1], cos_angle * arm[0] - arm_x[0]))
+        x_kin_range = min(x_kin_max - x_kin_min, x_kin_max * 2, -x_kin_min * 2)
+        print_width = config.getfloat('print_width', None, minval=0.,
+                                      maxval=x_kin_range)
+        if print_width:
```

Comment:
When print_width is defined, `min_angle` is effectively unused, right? Then I would suggest to prohibit the user to specify both options together, in order to avoid a false sense of security that both checks are active. E.g.

```
print_width = config.getfloat(...)
if print_width:
  self.limits[0] = (-print_width * 0.5, print_width * 0.5)
else:
  min_angle = config.getfloat('min_angle', ...)
  ...
```

And maybe reflect this in the documentation?

Reply:
min_angle also determine the maxval of print_width too. min_angle is only attended to protect against the risk of XZ inversion and the kinematics determines a safe x range from this parameter (specified or the value by default). If print_width is not specified the x range used is the "safe range". So my guess is that min_angle is a safety feature, print_width is a printer settings. The code already relate that. I would suggest adding a statement to the print_width documentation. Maybe:
> A safety limit is always determined using the min_angle value so that print_width cannot exceed it. If print_width is not specified, the valid toolhead X coordinates lie within these limits.

Reply:
Ah, I missed that relation. Then it makes sense, thanks!